src/main/scala/cognite/spark/v1/FlexibleDataModelsRelation.scala

```diff
@@ -537,27 +519,107 @@ class FlexibleDataModelsRelation(
+
+    i match {
+      case n: InstanceDefinition.NodeDefinition =>
+        ProjectedFlexibleDataModelInstance(
+          externalId = n.externalId,
+          properties = selectedInstanceProps.map {
+            case "externalId" => n.externalId
+            case p => allAvailablePropValues.get(p).map(extractInstancePropertyValue).orNull
+          }
+        )
+      case e: InstanceDefinition.EdgeDefinition =>
+        ProjectedFlexibleDataModelInstance(
+          externalId = e.externalId,
+          properties = selectedInstanceProps.map {
+            case "externalId" => e.externalId
+            case "startNode" => Array(e.startNode.space, e.startNode.externalId)
+            case "endNode" => Array(e.endNode.space, e.endNode.externalId)
+            case "type" => Array(e.relation.space, e.relation.externalId)
+            case p => allAvailablePropValues.get(p).map(extractInstancePropertyValue).orNull
+          }
+        )
+    }
+  }
 
-    ProjectedFlexibleDataModelInstance(
-      externalId = externalId,
-      properties = selectedInstanceProps.map { prop =>
-        allAvailablePropValues.get(prop).map(extractInstancePropertyValue).orNull
-      }
-    )
+  private def usageBasedFilterRequests(
+      limit: Option[Int],
+      instanceType: Option[InstanceType],
+      instanceFilter: Option[FilterDefinition]): Seq[InstanceFilterRequest] = {
+    val sources = Seq(InstanceSource(viewDefinition.toSourceReference))
+    val includeTyping = true
+    viewDefinition.usedFor match {
+      case Usage.Node =>
+        Seq(
+          InstanceFilterRequest(
+            instanceType = Some(instanceType.getOrElse(InstanceType.Node)),
+            filter = instanceFilter,
+            sort = None,
+            limit = limit,
+            cursor = None,
+            sources = Some(sources),
+            includeTyping = Some(includeTyping)
+          )
+        )
+      case Usage.Edge =>
+        Seq(
+          InstanceFilterRequest(
+            instanceType = Some(instanceType.getOrElse(InstanceType.Edge)),
+            filter = instanceFilter,
+            sort = None,
+            limit = limit,
+            cursor = None,
+            sources = Some(sources),
+            includeTyping = Some(includeTyping)
+          )
+        )
+      case Usage.All =>
+        Seq(
+          InstanceFilterRequest(
+            instanceType = Some(InstanceType.Node),
+            filter = instanceFilter,
+            sort = None,
+            limit = limit,
+            cursor = None,
+            sources = Some(sources),
+            includeTyping = Some(includeTyping)
+          ),
+          InstanceFilterRequest(
+            instanceType = Some(InstanceType.Edge),
+            filter = instanceFilter,
+            sort = None,
+            limit = limit,
+            cursor = None,
+            sources = Some(sources),
+            includeTyping = Some(includeTyping)
```

Comment:
This is a bit boiler plate, can just create a common filterRequest and then use `filterRequest.copy(instanceType= ...) `for this  and we can see directly what make the difference instead of scrolling on those duplicate line 😄